ur_robot_driver/src/hardware_interface.cpp

```diff
@@ -480,6 +544,64 @@ void HardwareInterface::read(const ros::Time& time, const ros::Duration& period)
     publishPose();
     publishRobotAndSafetyMode();
 
+    // Action feedback for joint trajectory forwarding
+    if (joint_forward_controller_running_)
+    {
+      control_msgs::FollowJointTrajectoryFeedback feedback = control_msgs::FollowJointTrajectoryFeedback();
+      for (size_t i = 0; i < 6; i++)
+      {
+        feedback.desired.positions.push_back(target_joint_positions_[i]);
+        feedback.desired.velocities.push_back(target_joint_velocities_[i]);
+        feedback.actual.positions.push_back(joint_positions_[i]);
+        feedback.actual.velocities.push_back(joint_velocities_[i]);
+        feedback.error.positions.push_back(std::abs(joint_positions_[i] - target_joint_positions_[i]));
+        feedback.error.velocities.push_back(std::abs(joint_velocities_[i] - target_joint_velocities_[i]));
+      }
+      jnt_traj_interface_.getHandle("joint_trajectory_handle").setFeedback(feedback);
+    }
```

Comment:
wouldn't the controller publish this?

Or is this done here so as to not have to change the `FollowJointTrajectory` controller?

Reply:
@gavanderhoorn The forwarding joint trajectory controller will take care of publishing feedback to the action client. Here, that feedback is filled with fresh inputs from the driver and passed back to the forwarding joint trajectory controller, which represents the primary action interface for the user. The original joint trajectory controller is not used in this scenario.

Reply:
So you're using the `FollowJointTrajectoryFeedback` message as a data structure, which gets exposed to controllers via an interface handle?

Reply:
Yes, that's right.